src/main/java/net/snowflake/client/core/SFSession.java

```diff
@@ -145,30 +144,51 @@ public QueryStatus getQueryStatus(String queryID) throws SQLException {
       statusUrl = sessionUrl + SF_PATH_QUERY_MONITOR + queryID;
     }
     // Create a new HTTP GET object and set appropriate headers
+
     HttpGet get = new HttpGet(statusUrl);
-    get.setHeader("Content-type", "application/json");
-    get.setHeader("Authorization", "Snowflake Token=\"" + this.sessionToken + "\"");
     String response = null;
     JsonNode jsonNode = null;
-    try {
-      response = HttpUtil.executeGeneralRequest(get, loginTimeout, getOCSPMode());
-      jsonNode = OBJECT_MAPPER.readTree(response);
-    } catch (Exception e) {
-      throw new SnowflakeSQLLoggedException(
-          this, e.getMessage(), "No response or invalid response from GET request. Error: {}");
-    }
-    // Get response as JSON and parse it to get the query status
-    // check the success field first
-    if (!jsonNode.path("success").asBoolean()) {
-      logger.debug("response = {}", response);
-
-      int errorCode = jsonNode.path("code").asInt();
-      throw new SnowflakeSQLException(
-          queryID,
-          jsonNode.path("message").asText(),
-          SqlState.SQLCLIENT_UNABLE_TO_ESTABLISH_SQLCONNECTION,
-          errorCode);
-    }
+    boolean sessionRenewed;
+    do {
+      sessionRenewed = false;
+      try {
+        get.setHeader("Content-type", "application/json");
+        get.setHeader("Authorization", "Snowflake Token=\"" + this.sessionToken + "\"");
+        response = HttpUtil.executeGeneralRequest(get, loginTimeout, getOCSPMode());
+        jsonNode = OBJECT_MAPPER.readTree(response);
+      } catch (Exception e) {
+        throw new SnowflakeSQLLoggedException(
+            this, e.getMessage(), "No response or invalid response from GET request. Error: {}");
+      }
+
+      // Get response as JSON and parse it to get the query status
+      // check the success field first
+      if (!jsonNode.path("success").asBoolean()) {
+        logger.debug("response = {}", response);
+
+        int errorCode = jsonNode.path("code").asInt();
+        if (errorCode == Constants.SESSION_EXPIRED_GS_CODE) {
+          try {
+            this.renewSession(this.sessionToken);
```

Comment:
Should we do something similar to https://github.com/snowflakedb/snowflake-jdbc/blob/c858fcdc229bc2e023dadc1837d30ce86a9916bf/src/main/java/net/snowflake/client/core/SFSession.java#L640-L673 to keep it thread safe?

Reply:
That calling of renewSession() is inside the heartbeat() function itself, meaning if the token expires during the call to refresh the token, we have to renew it first, so it's a special case. This case is more similar to https://github.com/snowflakedb/snowflake-jdbc/blob/master/src/main/java/net/snowflake/client/core/SFStatement.java#L479-L507, where the session is renewed following an expiration message from GS after a normal HTTP request. I modeled it off of this.

Reply:
Sorry I feel like didn't fully get the point. So to my current understanding, it's still possible that another thread using the same session (though not very likely) has already updated the token, namely issued and finished renewSession() before this thread issues renewSession()? But I guess it's only an optimization, current implementation can still guarantee the correctness.

Reply:
Since renewSession() is a `synchronized `function, wouldn't it be threadsafe anyway? I might not be understanding what you're suggesting as a change. Could you clarify?

Reply:
Support we have thread A and B. They share the same session. And both of got expired_gs_code at the same time. Then A called and finished renewSession() (updated the session token `this.sessionToken` as well) before B entered renewSession(). So when B entered renewSession(), it actually already has the valid session token. In this case, to my understanding, thread B doesn't need to update session token again.

Reply:
But even if my understanding is correct, the possibility of this case is still so rare. I am ok to not do it in this pr.